sympy/calculus/util.py

```diff
@@ -1631,3 +1632,161 @@ def union(self, other):
 
 # setting an alias for AccumulationBounds
 AccumBounds = AccumulationBounds
+
+def argmax_real(f, symbol = None, domain = S.Reals):
+    """
+    Returns the set of arguments of maxima, i.e., values of `symbol`
+    for which the `f` attains the maximum value in the given domain.
+    Note: This function uses derivatives to calculate the points where the
+    maximum value is achieved, and thus, would not include points where the
+    function is not differentiable. Also, the domain should be a subset of
+    S.Reals
+
+    Parameters
+    ==========
+
+    f: expr
+        The concerned expression
+    symbol: Symbol
+        Variable for which the arguments to be determined.
+    domain: Set
+        The domain over which the arguments are to be determined
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, sin, Interval, S, exp, pi
+    >>> from sympy.calculus.util import argmax_real
+
+    >>> x = Symbol('x')
+    >>> argmax_real(sin(x), x, Interval(-2*pi, pi))
+    FiniteSet(-3*pi/2, pi/2)
+
+    >>> argmax_real(-x**2 + 5*x - 6)
+    FiniteSet(5/2)
+
+    >>> argmax_real(exp(x), x, S.Reals)
+    EmptySet
+    """
+
+    return _argMaxMin(f, symbol , domain, max = True)
+
+def argmin_real(f, symbol = None, domain = S.Reals):
+    """
+    Returns the set of arguments of minima, i.e., values of `symbol`
+    for which the `f` attains the minimum value in the given domain.
+    Note: This function uses derivatives to calculate the points where the
+    minimum value is achieved, and thus, would not include points where the
+    function is not differentiable. Also, the domain should be a subset of
+    S.Reals
+
+    Parameters
+    ==========
+
+    f: expr
+        The concerned expression
+    symbol: Symbol
+        Variable for which the arguments to be determined.
+    domain: Set
+        The domain over which the arguments are to be determined
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, sin, Interval, S, exp, pi
+    >>> from sympy.calculus.util import argmin_real
+
+    >>> x = Symbol('x')
+    >>> argmin_real(sin(x), x, Interval(-2*pi, pi))
+    FiniteSet(-pi/2)
+
+    >>> argmin_real((x - 1)**2, x, S.Reals)
+    FiniteSet(1)
+
+    >>> argmin_real(exp(x), x, S.Reals)
+    EmptySet
+    """
+
+    return _argMaxMin(f, symbol , domain, max = False)
+
+def return_piecewise_solutions(f, x):
+    from sympy.logic.boolalg import And, Or
+
+    solutions = []
+    if type(f) == Piecewise:
```

Comment:
Maybe it would be better to use `piecewise_fold`

Reply:
@oscarbenjamin Yes that would be great. I didn't know this type of function existed. I would update the logic of this function that returns the suspected critical points in my next commit.